code/modules/chemistry/Chemistry-Dispenser.dm

```diff
@@ -271,224 +128,167 @@
 		qdel(src)
 		return
 
-	Topic(href, href_list)
-		if (status & (NOPOWER|BROKEN)) return
-		if (usr.stat || usr.restrained()) return
-		if (!in_range(src, usr)) return
-		if (isghostcritter(usr)) return
-		if (doing_a_thing) return
-
-		src.add_dialog(usr)
-		src.add_fingerprint(usr)
-
-		if (href_list["card"])
-			if (src.user_id)
-				src.eject_card()
-				src.update_account()
-				send_generic_details()
-				send_group_details()
-			else
-				var/obj/item/I = usr.equipped()
-				if (istype(I, /obj/item/card/id) || istype(I, /obj/item/card/data))
-					usr.drop_item()
-					I.set_loc(src)
-					src.user_id = I
-					src.update_account()
-					send_generic_details()
-					send_group_details()
-			return
-
-		else if (href_list["new_group"])
-			doing_a_thing = 1
-			var/reagents = input("Which reagents (separated by semicolons, indicate amount with equals signs)?","New Group") as null|text
-			if (isnull(reagents) || !length(reagents))
-				doing_a_thing = 0
-				return
-			var/name = input("What should the reagent group be called?","New Group") as null|text
-			name = copytext(sanitize(html_encode(name)), 1, MAX_MESSAGE_LEN)
-			if (isnull(name) || !length(name) || name == " ")
-				doing_a_thing = 0
-				return
-			//DEBUG(reagents)
-			var/list/reagentlist = params2list(reagents)
-			//DEBUG(list2params(reagentlist))
-
-			var/datum/reagent_group/G = new /datum/reagent_group()
-			for (var/reagent in reagentlist)
-				if (lowertext(reagent) in src.dispensable_reagents)
-					G.reagents += lowertext(reagent)
-					//Special amounts!
-					if (istext(reagentlist[reagent])) //Set a dispense amount
-						var/num = text2num(reagentlist[reagent])
-						if(!num) num = 10
-						G.reagents[lowertext(reagent)] = clamp(round(num), 1, 100)
-					else //Default to 10 if no specific amount given
-						G.reagents[lowertext(reagent)] = 10
-
-			if(G.reagents == 0)
-				doing_a_thing = 0
-				return
-
-			G.name = name
-			G.update_desc()
-			if (current_account)
-				current_account.groups += G
-			send_group_details()
-
-			doing_a_thing = 0
+	ui_interact(mob/user, datum/tgui/ui)
+		ui = tgui_process.try_update_ui(user, src, ui)
+		if(!ui)
+			ui = new(user, src, "ChemDispenser", src.name)
+			ui.open()
+
+	ui_static_data(mob/user)
+		var/list/static_data = list()
+		var/list/dispensableReagentsTemp = list()
+		if(dispensable_reagents)
+			for(var/reagent in dispensable_reagents)
+				var/datum/reagent/current_reagent = reagents_cache[reagent]
+				dispensableReagentsTemp.Add(list(list(
+					name = current_reagent.name,
+					colorR = current_reagent.fluid_r,
+					colorG = current_reagent.fluid_g,
+					colorB = current_reagent.fluid_b,
+					state = current_reagent.reagent_state,
+					id = reagent
+				)))
+		static_data["beakerName"] = glass_name
+		static_data["dispensableReagents"] = dispensableReagentsTemp
+
+		return static_data
+
+	ui_data(mob/user)
+		var/list/data = list()
```

Comment:
@ZeWaka Preference to use `.` for `data`?

Reply:
kinda, but it's not that hard to fix post merge